File: workflow/scripts/rename.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-

import os
import sys
import pandas as pd

# samples

def rename_fastqs(samplesheet, outdir):
    if not os.path.exists(outdir):
        os.makedirs(outdir)
    df = pd.read_csv(samplesheet, sep="\t", comment='#')
    df = df[df['Analysis'] == 1]
    for i, row in df.iterrows():
        f_id = row['FamilyID']
        s_id = row['SampleID']
        fq1 = row['FqR1']
        fq2 = row['FqR2']
        if pd.isna(fq1) and pd.isna(fq2):
            continue
        elif pd.isna(fq1):
            print(f"Error: {fq1} not found for sample {s_id}")
            sys.exit(1)
        elif pd.isna(fq2):
            assert fq1 is not None and os.path.exists(fq1)
            new_fq_path = os.path.join(outdir, f"{s_id}.fastq.gz")
            print(f"Symlinking {fq1} to {new_fq_path}")
            os.symlink(os.path.abspath(fq1), new_fq_path)
            df.loc[i, 'fq1'] = new_fq_path
        else:
            assert fq1 is not None and fq2 is not None and os.path.exists(fq1) and os.path.exists(fq2)
            new_fq1_path = os.path.join(outdir, f"{s_id}_R1.fastq.gz")
            new_fq2_path = os.path.join(outdir, f"{s_id}_R2.fastq.gz")
            print(f"Symlinking {fq1} to {new_fq1_path} and {fq2} to {new_fq2_path}")
            os.symlink(os.path.abspath(fq1), new_fq1_path)
            os.symlink(os.path.abspath(fq2), new_fq2_path)
            df.loc[i, 'fq1'] = new_fq1_path
            df.loc[i, 'fq2'] = new_fq2_path
    new_samplesheet = os.path.join(outdir, "samplesheet.csv")
    df.to_csv(new_samplesheet, sep="\t", index=False)


rename_fastqs(snakemake.params.samplesheet, snakemake.output.symlink_dir)

File: workflow/scripts/rename.R
rename_fastqs <- function(samplesheet, fastq_dir, out_dir) {
    sheet <- read.delim(samplesheet, sep="\t")

    ifelse(!dir.exists(out_dir), dir.create(file.path(out_dir), recursive = TRUE), FALSE)

    for (samp in sheet$sample) {
        print(paste("Renaming PE reads for:", samp))
        
        # Rename R1 reads
        file_1 <- paste0(fastq_dir, "/", unlist(strsplit(subset(sheet, sample == samp)$fq1, split = ",")))
        print(paste("Found", length(file_1), "R1 files for", samp))
        
        for (i in 1:length(file_1)) {
            if (file.exists(file_1[i])) {
                new_name <- paste0(out_dir, "/", samp, "-", i, "-R1.fastq.gz")
                system(paste0("ln -sr ", file_1[i], " ", new_name))
                print(paste(file_1[i], "successfully symlinked to", new_name))
            } else {
                stop(paste("Error:", file_1[i], "not found in", fastq_dir))
            }
        }
        
        # Rename R2 reads
        if (is.na(subset(sheet, sample == samp)$fq2) | subset(sheet, sample == samp)$fq2 == "") {
            print(paste("No R2 reads found for", samp))
            next
        }
        
        file_2 <- paste0(fastq_dir, "/", unlist(strsplit(subset(sheet, sample == samp)$fq2, split = ",")))
        print(paste("Found", length(file_2), "R2 files for", samp))
        
        for (i in 1:length(file_2)) {
            if (file.exists(file_2[i])) {
                new_name <- paste0(out_dir, "/", samp, "-", i, "-R2.fastq.gz")
                system(paste0("ln -sr ", file_2[i], " ", new_name))
                print(paste(file_2[i], "successfully symlinked to", new_name))
            } else {
                stop(paste("Error:", file_2[i], "not found in", fastq_dir))
            }
        }
    }
}

rename_fastqs(snakemake@params[["samplesheet"]], snakemake@params[["fastq_dir"]], snakemake@output[["symlink_dir"]])
